File: cpp/0069x的平方根.cpp
#include<iostream>

using namespace std;


// 二分
int mySqrt1(int x)
{
    int left = 0;
    int right = x;
    long mid;
    while (true)
    {
        mid = (left + right) >> 1;

        if(mid*mid==x)
            return mid;
        else if(mid*mid<x)
        {
            if((mid+1)*(mid+1)>x)
                return mid;
            else
                left = mid + 1;
        }

        else
            right = mid - 1;
    }
}

// 牛顿迭代
int mySqrt2(int x)
{
    if(x==0)
        return 0;
    double x1 = x;
    double x2 = x;
    while(true)
    {
        x2 = 0.5 * (x1 + x / x1);
        if(abs(x1-x2)<1e-1)
            return int(x2);
        x1 = x2;
    }
}

int main()
{
    int x = 2147395599;
    cout << mySqrt2(x);
}

File: python/main/0069x的平方根.py


# 二分
def mySqrt1(x):
    right=x
    left=0

    while(left<=right):
        mid=int((left+right)>>1)

        if(mid*mid==x):
            return mid
        elif(mid*mid<x):
            if((mid+1)*(mid+1)>x):
                return mid
            left=mid+1
        
        else:
            right=mid-1


# 牛顿迭代
def mySqrt2(x):
    if(x==0):
        return 0
    # 初值取为x
    x_prim,x_hat=x,x

    while(True):
        x_hat=(x_hat+x/x_hat)/2
        if(abs(x_hat-x_prim)<0.1):
            return int(x_hat)
        x_prim=x_hat
    
x=1364655888
print(mySqrt2(x))
        
        